hack/generator/pkg/astmodel/struct_type.go

```diff
@@ -127,5 +133,20 @@ func (structType *StructType) RelatedDefinitions(ref PackageReference, namehint
 		}
 	}
 
+	for name, function := range structType.functions {
+		funcDef := &FunctionDefinition{
+			Function:       function,
+			DefinitionName: DefinitionName{ref, name, &namehint},
+			Receiver:       &StructReference{DefinitionName{ref, namehint, nil}, false}} // TODO: This false right?
+		result = append(result, funcDef)
+	}
+
 	return result
 }
+
+// WithFunction attaches a function to the struct type
+func (structType *StructType) WithFunction(name string, function Function) *StructType {
+	structType.functions[name] = function
+
+	return structType
+}
```

Comment:
`StructType` is supposed to be immutable, so this function needs to return a copy of the struct, leaving the original unmodified. (This is a hedge against the same `StructType` (same `$ref`) being referenced with different names from different parts of the schema.)